### PR TITLE
fix_1416_iptc_DateCreated

### PR DESCRIPTION
**Pull request by clanmills**
See #1416.

### CODE REVIEW

**Comment by piponazo** [comment]
I was taking a look to the list of open issues and PRs, and I thought about continuing the work done here.

I noticed the following issue with the test file `tests/bash_tests/testcases.py` and test case 14:

```
        BT.log.info('#14 Convert a few other tags of interest from Exif/IPTC to XMP')
```

I extracted the minimum piece of code from that testcase into a bash script to reproduce the issue quickly:
```sh
#!/bin/bash
set -e

cp ../test/data/exiv2-empty.jpg /tmp/u.jpg
bin/exiv2 -M'set Exif.Photo.DateTimeOriginal 2003:12:14 12:01:44' /tmp/u.jpg
bin/exiv2 -M'set Exif.Photo.SubSecTimeOriginal 999999999' /tmp/u.jpg
bin/exiv2 -M'set Exif.Photo.ExifVersion 48 50 50 49' /tmp/u.jpg
bin/exiv2 -M'set Exif.Photo.ComponentsConfiguration 1 2 3 0' /tmp/u.jpg
bin/exiv2 -M'set Exif.Photo.Flash 73' /tmp/u.jpg
bin/exiv2 -M'set Exif.GPSInfo.GPSLatitude 3/1 8/1 29734512/1000000' /tmp/u.jpg
bin/exiv2 -M'set Exif.GPSInfo.GPSLatitudeRef N' /tmp/u.jpg
bin/exiv2 -M'set Exif.GPSInfo.GPSVersionID 2 2 0 1' /tmp/u.jpg
bin/exiv2 -M'set Exif.GPSInfo.GPSTimeStamp 1/1 2/1 999999999/1000000000' /tmp/u.jpg
bin/exiv2 -PEkycv /tmp/u.jpg
bin/exiv2 -pi /tmp/u.jpg
rm -f /tmp/u.xmp
echo "---- Extracting xmp"
bin/exiv2 -eX /tmp/u.jpg
echo "---- Printing xmp"
bin/exiv2 -px /tmp/u.xmp
```

When running this locally, I get:
```
(conan) luis@ryzenLinux:/media/linuxDev/programming/exiv2/buildRelease(fix_1416_iptc_DateCreated)$ ./script.sh 
Exif.Image.ExifTag                           Long        1  38
Exif.Photo.ExifVersion                       Undefined   4  48 50 50 49
Exif.Photo.DateTimeOriginal                  Ascii      20  2003:12:14 12:01:44
Exif.Photo.ComponentsConfiguration           Undefined   4  1 2 3 0
Exif.Photo.Flash                             Short       1  73
Exif.Photo.SubSecTimeOriginal                Ascii      10  999999999
Exif.Image.GPSTag                            Long        1  134
Exif.GPSInfo.GPSVersionID                    Byte        4  2 2 0 1
Exif.GPSInfo.GPSLatitudeRef                  Ascii       2  N
Exif.GPSInfo.GPSLatitude                     Rational    3  3/1 8/1 29734512/1000000
Exif.GPSInfo.GPSTimeStamp                    Rational    3  1/1 2/1 999999999/1000000000
---- Extracting xmp
---- Printing xmp
Warning: Unsupported date format  <--------------- This is the warning causing problems with the test
Xmp.exif.ExifVersion                         XmpText     4  2.21
Xmp.exif.GPSVersionID                        XmpText     7  2.2.0.1
Xmp.exif.GPSLatitude                         XmpText    12  3,8.4955752N
Xmp.exif.GPSTimeStamp                        XmpText    29  2003-12-14T01:02:00.999999999
Xmp.exif.ComponentsConfiguration             XmpSeq      4  YCbCr
Xmp.exif.Flash                               XmpText     0  type="Struct"
Xmp.exif.Flash/exif:Fired                    XmpText     4  True
Xmp.exif.Flash/exif:Return                   XmpText     1  0
Xmp.exif.Flash/exif:Mode                     XmpText     1  1
Xmp.exif.Flash/exif:Function                 XmpText     5  False
Xmp.exif.Flash/exif:RedEyeMode               XmpText     4  True
Xmp.photoshop.DateCreated                    XmpText    29  2003-12-14T12:01:44.999999999
```

And the content of the file u.xmp is:
```
<?xpacket begin="﻿" id="W5M0MpCehiHzreSzNTczkc9d"?>
<x:xmpmeta xmlns:x="adobe:ns:meta/" x:xmptk="XMP Core 4.4.0-Exiv2">
 <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
  <rdf:Description rdf:about=""
    xmlns:exif="http://ns.adobe.com/exif/1.0/"
    xmlns:photoshop="http://ns.adobe.com/photoshop/1.0/"
   exif:ExifVersion="0221"
   exif:GPSVersionID="2.2.0.1"
   exif:GPSLatitude="3,8.4955752N"
   exif:GPSTimeStamp="2003-12-14T01:02:00.999999999"
   photoshop:DateCreated="2003-12-14T12:01:44.999999999">
   <exif:ComponentsConfiguration>
    <rdf:Seq>
     <rdf:li>1</rdf:li>
     <rdf:li>2</rdf:li>
     <rdf:li>3</rdf:li>
     <rdf:li>0</rdf:li>
    </rdf:Seq>
   </exif:ComponentsConfiguration>
   <exif:Flash
    exif:Fired="True"
    exif:Return="0"
    exif:Mode="1"
    exif:Function="False"
    exif:RedEyeMode="True"/>
  </rdf:Description>
 </rdf:RDF>
</x:xmpmeta>
```
Note that the `exif:GPSTimeStamp` also gets wrong in the process of generating the XMP file.

I'll keep investigating this issue.

**Comment by clanmills** [comment]
I think the message `Unsupported date format` (and `Unsupported time format`) are written to errour-out by the runtime library.  In the good old days when bash_tests were written in bash, I think we piped error-out to /dev/null hide this _( `2>/dev/null`)_.  When bash_tests were rewritten in python, the pipe has been lost.

**Comment by clanmills** [comment]
@piponazo I haven't reproduced this on Ubuntu and macOS. 

Which platform are you using?

BTW, I think you need `export "LD_LIBRARY_PATH=$PWD/lib:$PWD"` to get bin/exiv2 to use the local library in build/lib.

**Comment by piponazo** [comment]
> Which platform are you using?

I'm using Ubuntu 20.04 and gcc 9.3. I am quite confident that the **exiv2** binary I am using is pointing to the libexiv2 library I am building locally. I could actually run the debugger and go step by step through the code.

```bash
(conan) luis@ryzenLinux:/media/linuxDev/programming/exiv2/buildRelease(fix_1416_iptc_DateCreated)$ ldd bin/exiv2
	linux-vdso.so.1 (0x00007ffce2bd7000)
	libexiv2.so.0 => /media/linuxDev/programming/exiv2/buildRelease/lib/libexiv2.so.0 (0x00007f3eb918b000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f3eb8f8f000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f3eb8f74000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f3eb8d82000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f3eb8c33000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f3eb95af000)
```

My major concern at the moment is about the drift of values in the XMP file:

```
   exif:GPSLatitude="3,8.4955752N"
   exif:GPSTimeStamp="2003-12-14T01:02:00.999999999"
```
I need to check if this is happening on `main` or `0.27-maintenance` already, or only in this branch.

**Comment by clanmills** [comment]
Right.  I'm using '0.27-maintenance'.  I'm wrong about this coming from the runtime library.  I think it's coming from the src/value.cpp.  Let me get onto "the same page".

**Comment by piponazo** [comment]
I have finally found why this is happening. When reading the XMP, in the file `convert.cpp:1153` we have the following variable values:

```
to: "Iptc.Application2.DateCreated"
value: "2003-12-14T12:01:44.999999999"
```

Before the precondition of the function `DateValue::read` was

```cpp
int DateValue::read(const std::string& buf)
    {
        // Hard coded to read Iptc style dates
        if (buf.length() < 8) {
    ...
```

And now the condition is: `    if (buf.length() != 8 && buf.length() != 10) {`. That's why we are getting the warning message now. 

I will try to first fix this issue quickly by massaging the condition. However, it might be a better idea to use std::regex to just extract the date from the whole string.

**Comment by clanmills** [comment]
Good idea to use std::regex.  I'm sure this code was written 20 years ago when std::regex lay in the future.

I don't know the rational for the "convertor" code.  Life would be simpler if exiv2 were deterministic. It should report, write and edit metadata exactly as requested by the user.  I have not documented the "convertor" code in my book because I haven't understood its purpose.

**Comment by piponazo** [comment]
Some of the other failing tests where complaining about expectations in the date format:

```
1: [INFO] diff_bytes:
1: <<< /media/linuxDev/programming/exiv2/test/data/glider.exv.iagd
1: >>> glider.exv.iatst
1: 0,0c1
1: ---
1: > Warning: Unsupported date format
1: 6c7
1: < Iptc.Application2.DateCreated                0x0037 Date        8  2004-08-03
1: ---
1: > Iptc.Application2.DateCreated                0x0037 String      8  2004-8-3
1: [INFO] diff_bytes:
1: <<< /media/linuxDev/programming/exiv2/test/data/iptc-noAPP13.jpg.iagd
1: >>> iptc-noAPP13.jpg.iatst
1: 0,0c1
1: ---
1: > Warning: Unsupported date format
1: 5c6
1: < Iptc.Application2.DateCreated                0x0037 Date        8  2004-08-03
1: ---
1: > Iptc.Application2.DateCreated                0x0037 String      8  2004-8-3
```
In one of our python utils file, we had something like this:

```
m Iptc.Application2.DateCreated		  2004-8-3
```

With I have changed to use the ISO_8601 date format:
```
m Iptc.Application2.DateCreated		  2004-08-03
```

and it seems like no more tests are failing after than change.

Whoever, it seems like we are accepting at the moment Date formats like `2004-8-3`. Do you think we should be more strict here and accept only dates with the ISO format `YYYY-MM-DD`?

**Comment by piponazo** [comment]
In the original issue @Jehan refers to the page https://exiv2.org/iptc.html where it is written the following: 

> 0x0037 | 55 | Iptc.Application2.DateCreated | Date | No | No | 8 | 8 | Represented in the form CCYYMMDD to designate the date the intellectual content of the object data was created rather than the date of the creation of the physical representation. Follows ISO 8601 standard.

From https://en.wikipedia.org/wiki/ISO_8601 I extracted this:

>Representations can be done in one of two formats – a basic format with a minimal number of separators or an extended format with separators added to enhance human readability.[14][15] The standard notes that "The basic format should be avoided in plain text."[16] The separator used between date values (year, month, week, and day) is the hyphen, while the colon is used as the separator between time values (hours, minutes, and seconds). For example, the 6th day of the 1st month of the year 2009 may be written as "2009-01-06" in the extended format or simply as "20090106" in the basic format without ambiguity.

And actually from the ISO itself ([link](https://web.archive.org/web/20171020084445/https://www.loc.gov/standards/datetime/ISO_DIS%208601-1.pdf) we can see that both: basic and extended formats are considered (section 4.1.3.2):

```
Basic format: YYYYDDD Example: 1985102
Extended format: YYYY-DDD Example: 1985-102
```

I will play a bit with regular expressions and try to get this right.

**Comment by clanmills** [comment]
Well, the first goal is to make the test suite pass.  Anything for a quiet life!

What date formats should be accepted?  I think we'd have to look at the standard.  I'm believe Xmp uses ISO standards because it's based on XML technology.  Exif uses it's own 20 ascii byte format:  `YYYY:MM:DD hh:mm:ss\0`  @postscript-dev has been working with the IPTC folks.  Peter do you know anything about IPTC date/time formats?

```bash
739 rmills@rmillsm1:~/gnu/github/exiv2/0.27-maintenance $ exiv2 -g date/i ~/Stonehenge.jpg 
Exif.Image.DateTime                          Ascii      20  2015:07:16 20:25:28
Exif.Photo.DateTimeOriginal                  Ascii      20  2015:07:16 15:38:54
Exif.Photo.DateTimeDigitized                 Ascii      20  2015:07:16 15:38:54
Exif.NikonWt.DateDisplayFormat               Byte        1  Y/M/D
Exif.GPSInfo.GPSDateStamp                    Ascii      11  2015:07:16
Xmp.xmp.ModifyDate                           XmpText    25  2015-07-16T20:25:28+01:00
740 rmills@rmillsm1:~/gnu/github/exiv2/0.27-maintenance $ 
```

@piponazo We're going out now and home in about 4 hours (about 15:00 UTC).

**Comment by clanmills** [comment]
Here's the IPTC standard:
```bash
741 rmills@rmillsm1:~/gnu/github/exiv2/0.27-maintenance $ curl --silent -L https://iptc.org/std/photometadata/specification/iptc-pmd-techreference_2021.1.json | grep -ie time -ie date
  "document_timestamp": "2021-10-20T18:59:50+00:00",
    "dateCreated": {
      "name": "Date Created",
      "specidx": "#date-created",
      "dataformat": "date-time",
 ...
 ```
 And https://www.iptc.org/std/photometadata/specification/IPTC-PhotoMetadata
 
 And here's the spec for date _(which seems to also deal with time)_.
 
 https://www.iptc.org/std/photometadata/specification/IPTC-PhotoMetadata#date-value-type

**Comment by postscript-dev** [comment]
I added a section to the manpage to try and explain the date/time formats ([7.6 Date/Time formats](https://github.com/Exiv2/exiv2/blob/fd8447129cc8c8c5dd97fcaa1af0254f7494e04e/exiv2.md#date_time_fmts)), but for some reason didn't include XMP. Unless someone else wants to add this, I will include it in my list of corrections/updates for the manpage.

File: src/value.cpp
Comment (kevinbackhouse) [inline src/value.cpp:1026]:
```suggestion
        char b[15] = {};
```

File: src/value.cpp
Comment (kevinbackhouse) [inline src/value.cpp:1027]:
```suggestion
        std::memcpy(b, reinterpret_cast<const char*>(buf), std::min(len, sizeof(b)-1));
```

File: src/value.cpp
Comment (piponazo) [reply src/value.cpp]: I have replaced that code by the usage of regular expressions. I added few more unit tests to be sure we are supporting the different formats mentioned in the ISO.

**Comment by piponazo** [comment]
I think the PR is ready for review now :wink:

**Review by clanmills — COMMENTED** [review]
Thanks for dealing with this, @piponazo.  The part of this that'll make the warning disappear are the changes to the test code.  However, introducing the reg-ex for ISO and ISO-Ext dates is a great idea.  You do more with less code.

I don't think this can be pack ported to branch 0.27-maintenance as it requires C++11 features such as `R"literal"`.  So be it.  These message have been in 0.27-maintenance for years.  I think the code in `$ make bash_tests` piped the message into /dev/null.  It's not worth spending time on this for 0.27-maintenance.

I can't approve this PR because I am the author.  Either somebody else can approve this, or submit a new PR for me to approve and I'll close this (and delete the associated branch).

**Comment by piponazo** [comment]
Thanks @clanmills for your feedback. Once somebody else review and approve this PR, I might also spend some time to port the minimum stuff to 0.27. However, as you mentioned, with the regular expressions (which is not available in C++98) the code gets simpler and cover more cases. For 0.27, I would just try to cover the most common cases (and fix the test code).

**Comment by clanmills** [comment]
@piponazo Don't bother touching the C++ code that causes "unsupported date format" to be issued.  Just modify the test harness.  The simplistic code that generates these messages has been in Exiv2 forever.

You've made the code much better on 'main' and 0.27-maintenance doesn't deserve more attention.  You could even introduce new bugs.  Best to leave it alone!

File: src/value.cpp
Comment (kmilos) [inline src/value.cpp:978]: If so, then missing padding w/ 0 to width 4 for they date_.year below?

File: src/value.cpp
Comment (kmilos) [inline src/value.cpp:919]: As discussed [here](https://github.com/Exiv2/exiv2/pull/1955), there is a possibility to support years before BC in ISO 8601 (unless IPTC expressly prohibits these), so an optional [+-] prefix for the year could be included as well?

Comment (piponazo) [reply src/value.cpp:919]: According to the ISO document, only values from 0000 to 9999 should be considered:

> calendar year is, unless specified otherwise, represented by four digits. Calendar years are
numbered in ascending order according to the Gregorian calendar by values in the range [0000]
to [9999]. Values in the range [0000] through [1582] shall only be used by mutual agreement of
the partners in information interchange.

I actually took these regular expressions from the book "Regular Expressions Cookbook":
https://learning.oreilly.com/library/view/regular-expressions-cookbook/9781449327453/ch04s04.html

File: src/value.cpp
Comment (piponazo) [reply src/value.cpp:978]: Good catch. I will add the padding although as it is mentioned in the ISO:

> Values in the range [0000] through [1582] shall only be used by mutual agreement of
the partners in information interchange.

Note that I did not change that code. Probably most of the dates handled by Exiv2 will be between 19xx & 20XX

File: src/value.cpp
Comment (kmilos) [reply src/value.cpp:919]: The standard does allow for negative years in the "expanded form" (to be agreed between exchanging parties), [as mentioned on Wikipedia](https://en.wikipedia.org/wiki/ISO_8601#Years). Just don't know if that's applicable here or not, not that familiar w/ IPTC...

Comment (piponazo) [reply src/value.cpp:919]: From your wikipedia link:

> SO 8601 prescribes, as a minimum, a four-digit year [YYYY] to avoid the year 2000 problem. It therefore represents years from 0000 to 9999, year 0000 being equal to 1 BC and all others AD. However, years before 1583 are not automatically allowed by the standard. Instead "values in the range [0000] through [1582] shall only be used by mutual agreement of the partners in information interchange."[19]. To represent years before 0000 or after 9999, the standard also permits the expansion of the year representation but only by prior agreement between the sender and the receiver.[20] 

IMHO, I think anything below 1900 is not realistic for representing Creation/Modification dates in digital images. I would not add the optional [+|-] unless more of you guys think it would cover some valid cases.

**Comment by clanmills** [comment]
I think what you have done is much better than it used to be.  If a user wants BC dates _(for archeological purposes perhaps?)_, they can submit an issue report with a use case.

**Comment by clanmills** [comment]
@kmilos I'll leave it for you to approve this change when you are happy.

**Comment by kmilos** [comment]
Exactly, I imagine this is useful only to the digital archives of the National Museum of Mesopotamia and such...

So we could indeed just wait for a bug report from them if the others are ok with that.

**Comment by piponazo** [comment]
I just added the support for the optional [+|-] and I am starting to regret about it ... I thought about adding the following additional unit tests and it provokes a crash:

```cpp
TEST(ADateValue, copiesToByteBufferWithBasicFormat_negativeYear)
{
    const DateValue dateValue (-2021, 12, 1);
    std::array<byte, 8> buf;
    buf.fill(0);

    const byte expectedDate[8] = {'2', '0', '2', '1', '1', '2', '0', '1'};
    ASSERT_EQ(8, dateValue.copy(buf.data()));
    ASSERT_TRUE(std::equal(buf.begin(), buf.end(), expectedDate));
}
```

The `DateValue::copy` method is exporting the date in the basic format YYYYMMDD with exactly 8 characters:

```
        char temp[9];
        int wrote = snprintf(temp, sizeof(temp), "%04d%02d%02d", date_.year, date_.month, date_.day);
        assert(wrote == 8);
```

See also https://exiv2.org/iptc.html.

We should probably just support positive years by design to avoid issues with different date formats in different tag standards.

If we agree on this, I would drop the last commit I pushed few minutes ago.

**Comment by kmilos** [comment]
> We should probably just support positive years by design to avoid issues with different date formats in different tag standards.
> 
> If we agree on this, I would drop the last commit I pushed few minutes ago.

If this was fixing a specific IPTC schema issue, and we're not taking away any accidentally present negative year functionality, then by all means.

**Review by kmilos — APPROVED** [review]

**Comment by Jehan** [comment]
Thanks all for working on this. Sorry I haven't had the time myself to complete my patch (between GIMP and other stuff, I missed the time). Hopefully we'll soon see this issue fixed then! 😃

**Comment by clanmills** [comment]
Thanks @Jehan for taking the time to say thanks.  It is appreciated.  And thank you for your work on GIMP.

Well done to everybody in Team Exiv2 for working on this.  The addition of the std::regex code is elegant and the observation about BC dates shows attention to detail.  After years of lonely efforts on Exiv2, it's great to see the library (and the documentation) moving forward in such a positive manner.  Well done, everybody.

File: src/value.cpp
Comment (jim-easterbrook) [reply src/value.cpp:919]: Scans of 19th century photographs can have creation dates prior to 1900.  (:

**Pull request merged**
